### PR TITLE
MAINT: Add ci skip to azp

### PR DESCRIPTION
**Pull request by larsoner**
Over in MNE we've been using a two-stage Azure pipeline for 6-12 months and so far it has worked:

1. Check for `[ci skip]`, `[skip ci]`, `[skip azp]`, `[azp skip]` in the commit message
2. If found, don't run

This should help with things like https://github.com/scipy/scipy/pull/13810 where we don't want Azure to run.

We can also do something similar with GH-actions if we want finer-grained control like `[skip github]` to work, but maybe we don't need it for now. It would be nice to at least have `[ci skip]` skip all CIs.

The first commit here has `[ci skip]` in the message, so in theory we should be able to see it in action.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -30,313 +30,335 @@ variables:
...
+          git_log=`git log --max-count=1 --skip=1 --pretty=format:"%s"`
+          echo "##vso[task.setvariable variable=log]$git_log"
+      - bash: echo "##vso[task.setvariable variable=RET]false"
+        condition: or(contains(variables.log, '[skip azp]'), contains(variables.log, '[azp skip]'), contains(variables.log, '[skip ci]'), contains(variables.log, '[ci skip]'))
+      - bash: echo "##vso[task.setvariable variable=start_main;isOutput=true]$RET"
+        name: result
+
+- stage: Main
+  condition: and(succeeded(), eq(dependencies.Check.outputs['Skip.result.start_main'], 'true'))
+  dependsOn: Check
```
Comment (larsoner) [inline azure-pipelines.yml:52]: The diff looks terrible, but everything other than this green block is unchanged other than being indented an extra two spaces.

**Comment by tupui** [comment]
Should we have something more precise in term of message? I would love to have an easy way to do things like `skip doc` or `only doc` (same for the rest).

**Comment by larsoner** [comment]
> Should we have something more precise in term of message? I would love to have an easy way to do things like skip doc or only doc (same for the rest).

The standard way AFAIK is to use a blocklist like implemented here, rather than what I think you're asking for, which is essentially an allowlist. I'd rather not deviate from it.

For example, `[skip ci]` and `[ci skip]` are respected by most/all other CIs that I've worked with and Azure is just behind the times for those. Adding the more targeted `[skip azp]` and `[azp skip]` are motivated by the historical fact that their equivalents were recognized by Travis and AppVeyor (back when we used them, and others still do), and there are issues on the forums for other CIs (GitHub actions, CircleCI) where people request these. So there is an acquired joint community knowledge of this way of doing things. So to get a doc-only build for example at one point the standard way to do was to use a blocklist as `[skip travis] [skip appveyor]` and let CircleCI be the only thing that ran. And we recently did something similar [for GitHub actions](https://github.com/scipy/scipy/blob/master/.github/workflows/linux.yml#L54) (though I'm not 100% sure it works, and if it doesn't, can be easily fixed using another build step).

I also think that this blocklist is actually better in principle than an allowlist. If we went to an allowlist-based system, any time you want to add a new CI mode, you have to refactor every CI to look for any of the allow-based keywords, which creates dependencies between CIs that otherwise do not need to be there.

**Comment by tupui** [comment]
Thanks for all the details. I understand, makes sense.

**Review by rgommers — APPROVED** [review]
Great idea, thanks @larsoner.

There's this Azure issue I find every year or so with a ton of people complaining about not respecting `[ci skip]` on PRs, even though it does on merges to master. This is a nice way of working around that.

**Comment by rgommers** [comment]
Merged, thanks @larsoner

**Pull request merged**